Allow configuring http connection pool size

kdvolder opened the pull request:
Adresses https://github.com/docker-java/docker-java/issues/1466

bsideup commented on docker-java-api/pom.xml:
```diff
@@ -4,7 +4,7 @@
 		<groupId>com.github.docker-java</groupId>
 		<artifactId>docker-java-parent</artifactId>
-		<version>3.2.2-SNAPSHOT</version>
+		<version>3.2.6-SNAPSHOT</version>
```
could you please revert the version change?

kdvolder replied:
I can, but why? It actually makes it harder for me to do local build of this and test it against STS (which is where we use it). The wrong version will mess up my local setup that consumes the dependencies.

I would suggest that its better to strive for an accurate version number in there. I know you override the version at build-time so it doesn't matter what you put in there (so again, makes me wonder why you'd care if I 'fix it' to be accurate. Why would you insist on keeping an old version in there? There seems to be no benefit to that whatsoever).

bsideup replied:
> The wrong version will mess up my local setup that consumes the dependencies.
> insist on keeping an old version in there

soon, there will be no "current" version at all, just `0-SNAPSHOT`. So I am insisting on not touching the version at all, because this is not feature PR's responsibility.

> makes me wonder why you'd care if I 'fix it' to be accurate

Even if you're willing to "fix it", that would be a separate PR that changes it to `0-SNAPSHOT`, according to the plan.

kdvolder replied:
Okay. I will do as you wish of course. But please allow me to address this comment as the very last one. 

I.e. once everything else about this PR is acceptable, then I will address this comment. If I were to do it now, it just adds a bunch friction around the local process I'm currently needing to follow to validate that the PR actually works for us.

kdvolder replied:
Change reverted

bsideup commented on docker-java-transport-httpclient5/src/main/java/com/github/dockerjava/httpclient5/ApacheDockerHttpClient.java:
```diff
@@ -23,13 +25,18 @@ public Builder sslConfig(SSLConfig value) {
             return this;
         }
 
+        public Builder connectionPool(ConnectionPoolConfig conf) {
```
Since in most cases the route is always the same, I recommend we simplify this to just `Integer connectionPoolSize`

kdvolder replied:
That would work for our use case so... sure. I just thought it might be more logical to just copy the underlying config options exactly. It is also slightly more future proof. Don't really feel strongly about it either way so I'll do as you ask.

kdvolder replied:
Done!

bsideup commented on docker-java-transport-httpclient5/src/main/java/com/github/dockerjava/httpclient5/ApacheDockerHttpClientImpl.java:
```diff
@@ -66,27 +66,36 @@ protected ApacheDockerHttpClientImpl(
                 host = HttpHost.create(dockerHost);
         }
 
+        PoolingHttpClientConnectionManager connectionPoolMgr = new PoolingHttpClientConnectionManager(
```
please keep the variable naming consistent with the rest of the library

kdvolder replied:
Please be specific, I cannot guess easily what you want. Tell me what name you like to see and I'll change it.

bsideup replied:
`connectionManager` would be fine, thanks

kdvolder replied:
Done!

bsideup reviewed (commented):
Thanks for working on this!

Before accepted, this change could use some tests :)

kdvolder commented:
> Before accepted, this change could use some tests :)

Sure! But I'm not seeing a 'src/test' folder in either of the projects that I made changes to. So I'm not really sure where tests for these projects supposed to live.

So please provide some guidance. E.g. where should I place a test, and how do you run them typically (so I can give em trial run before committing / pushing).

Thanks for your help!

bsideup commented:
@kdvolder all legacy tests are in `docker-java/src/test/main` but it would be better to put this one into module's test sources. Please let me know if you feel that you need help with that - I can take over the PR and assist you

kdvolder commented:
For the time being I'm making small commits for addressing individual comments. Once the PR is 'acceptable' and ready to merge I can rebase and squash into a single commit... or not... depending on what you prefer.

kdvolder commented:
I'm looking at adding a test or two but I'm not really sure how to do it, or even what kind of test to add.

I think it would make sense to add a basic test that checks that:

- if you create a zerodep/appache http client instance with a custom 'ConnectionPoolConfig'
- then you get a client that has the expected connection pool manager setup (i.e a org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager.PoolingHttpClientConnectionManager instance with the right characteristics.

Problem that I am having is that the apis are not really letting me access internal details such as `PoolingHttpClientConnectionManager`. So the test can't get access to this instance to check it is setup as expected.

I could try adding more methods to the api so as to expose more of this the internals publically. Byt probably you wouldn't like it, because the feel I get from the api is that it is deliberately trying to not expose this sort of information on purpose.

So if you'd like me write some tests, which I'm happy to do, I'm going to need some concrete suggestions on what exactly to do.

Not sure if guiding me through this may actually be more work for you then writing a simple test case yourself. 

I'd be happy for your guidance but if you rather take it from here and write the test yourself, I'm fine with that as well. Whatever is more convenient to you.

kdvolder commented:
@bsideup I think I have addressed all comments except for adding tests. 

I don't know what to do about the tests and will leave that in your capable hands as you have so kindly offered:

> Please let me know if you feel that you need help with that - I can take over the PR and assist you

In the mean time I have already built jars for our use from the forked repo, and we have adopted the use of those jars for the time being. As far as I am concerned that resolves the issue on our end. Though I would really prefer if this could be merged into master so we do not depend on a custom forked build. If you want me to add tests, I am happy to do so, but I can't do it without your help.

kdvolder commented:
Squashed into a single commit

bsideup commented:
@kdvolder I just merged your PR into a branch and will take it from there. Thanks for working on it 👍

kdvolder commented:
@bsideup Thanks for taking care of the test case and final polishes.
